odl/solvers/advanced/chambolle_pock.py

```diff
@@ -0,0 +1,278 @@
+    Kadjoint = K.adjoint
+
+    for nn in range(niter):
+        # Copy required for relaxation
+        xold = x.copy()
+        # Gradient descent in the dual variable y
+        f_cc_prox_sigma(y + sigma * K(xbar), out=y)
+        # Gradient descent in the primal variable x
+        g_prox_tau(x - tau * Kadjoint(y), out=x)
+        # Overrelaxation in the primal variable x
+        xbar = x + theta * (x - xold)
+
+        if partial is not None:
+            partial.send(x)
+
+    return x
+
+
+def f_cc_prox_l2_tv(space, g, lam):
+    """Function for the proximal operator with l2-data plus TV-regularization.
+
+    Factory function which provides a function to initialize the proximal
+    operator of the convex conjugate of the functional ``F`` given by the
+    L2-data term and the isotropic total variation semi-norm as in the primal
+    minimization problem:
+
+    F(y,z)= 1/2 * ||y - g||_2^2 + lambda * ||(|grad u|)||_1
+
+    with y = Au and z = grad u. The operators ``A`` and ``grad`` are
+    combined to a matrix operator ``K`` as:
+
+        K = (A, grad)^T
+
+    The proximal operator is mapping from a vector space X to its dual space
+    X^*. We assume X to be Hilbert spaces which is self-dual. Here, the domain
+    and range of the proximal operator are given by the range of ``K``.
+
+    Parameters
+    ----------
+    space : `ProductSpace`
+        Product space of the range of forward operator ``A`` and of the range
+        of the gradient operator i.e. a product space of the image space
+    g : `DiscreteLpVector`
+        Element in the range of the forward operator ``A``
+    lam : positive `float`
+        Regularization parameter
+
+    Returns
+    -------
+    make_prox : `function`
+        Function initialize the proximal operator with a given step
+        length parameter ``sigma``
+    """
+
+    def make_prox(sigma):
+        """Returns an instance of the proximal operator.
+
+        Parameters
+        ----------
+        sigma : positive `float'
+            Step length parameter
+
+        Returns
+        -------
+        prox_op : `Operator`
+            Proximal operator initialized with ``sigma``
+        """
+
+        class _prox_op(Operator):
+
+            """The proximal operator."""
+
+            def __init__(self, sigma):
+                """Initialize the proximal operator.
+
+                Parameters
+                ----------
+                sigma : positive `float`
+                """
+                self.sigma = sigma
+                super().__init__(domain=space, range=space)
+
+            def _call(self, x, out):
+                """Apply the operator to ``x`` and stores the result in
+                ``out``"""
+
+                y = x[0]
+                z = x[1]
+
+                # First component: (y - sig*g) / (1 + sig)
+
+                sig = self.sigma
+                out[0][...] = y / sig
+                out[0] -= g
+                out[0] *= sig / (1 + sig)
```

Comment:
Second thought, this whole thing could be 1 lincomb:

```python
out[0].lincomb(1 / (1 + sig), y, -sig / (1 + sig), g)
```

significantly faster and briefer code.